File: src/com/aslan/module/cipher/keys/KeyFactoryProxy.js
import KeyFactory from "../KeyFactory.js";

export default class KeyFactoryProxy extends KeyFactory {
    level;
    constructor(level) {
        super();
        this.level = level;
    }

    make(option) {
        let version = option.levelV || this.level.version[0].id;
        let find = this.level.version.find(e => e.id === version);
        if (find == null) {
            throw new Error("不支持的密匙算法版本({@name})：{}".format(this.levelV, version));
        }
        let Class = require(find.class);
        option.levelV = version;
        return new Class.default();
    }
}

File: src/com/aslan/module/cipher/keys/KeyFactoryProxy.java
package com.aslan.module.cipher.keys;

import com.aslan.module.cipher.CipherInfo;
import com.aslan.module.cipher.Key;
import com.aslan.module.cipher.KeyFactory;
import com.aslan.module.core.DiffusionException;

import java.util.SortedMap;
import java.util.TreeMap;


public abstract class KeyFactoryProxy implements KeyFactory {
    protected SortedMap<Integer, Class<? extends Key>> map = new TreeMap<>();

    public Key make(CipherInfo option) throws Exception {
        int version = option.levelV;
        Class<? extends Key> cl = form(version);
        if (cl == null) {
            throw new DiffusionException("不支持的密匙算法版本({@name})：{}");
        }
        Key key = cl.newInstance();
        option.levelV = key.version();
        return key;
    }

    protected Class<? extends Key> form(int version) {
        if (version == 0) version = map.firstKey();
        return map.get(version);
    }
}